lib/ncurses/tack/tack.1
.\"***************************************************************************
.\" Copyright (c) 1998-2005,2006 Free Software Foundation, Inc.              *
.\"                                                                          *
.\" Permission is hereby granted, free of charge, to any person obtaining a  *
.\" copy of this software and associated documentation files (the            *
.\" "Software"), to deal in the Software without restriction, including      *
.\" without limitation the rights to use, copy, modify, merge, publish,      *
.\" distribute, distribute with modifications, sublicense, and/or sell       *
.\" copies of the Software, and to permit persons to whom the Software is    *
.\" furnished to do so, subject to the following conditions:                 *
.\"                                                                          *
.\" The above copyright notice and this permission notice shall be included  *
.\" in all copies or substantial portions of the Software.                   *
.\"                                                                          *
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS  *
.\" OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF               *
.\" MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.   *
.\" IN NO EVENT SHALL THE ABOVE COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,   *
.\" DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR    *
.\" OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR    *
.\" THE USE OR OTHER DEALINGS IN THE SOFTWARE.                               *
.\"                                                                          *
.\" Except as contained in this notice, the name(s) of the above copyright   *
.\" holders shall not be used in advertising or otherwise to promote the     *
.\" sale, use or other dealings in this Software without prior written       *
.\" authorization.                                                           *
.\"***************************************************************************
.\"
.\" $Id$
.TH tack 1M ""
.ds n 5
.ds d @TERMINFO@
.SH NAME
\fBtack\fR - \fIterminfo\fR action checker
.SH SYNOPSIS
\fBtack\fR [-itV] [term]
.br
.SH DESCRIPTION
The \fBtack\fR program has three purposes:
(1) to help you build a new terminfo entry describing an unknown terminal,
(2) to test the correctness of an existing entry, and
(3) to develop the correct pad timings needed to ensure that screen updates
don't fall behind the incoming data stream.
.PP
\fBTack\fR presents a series of screen-painting and interactive
tests in ways which are intended to make any mismatches between the
terminfo entry and reality visually obvious. 
\fBTack\fR also provides tools that can help in understanding how
the terminal operates.
.SS OPTIONS
.TP
.I "\-i"
Usually \fBtack\fR will send the reset and init strings to the terminal
when the program starts up.  The \fI-i\fR option will inhibit the
terminal initialization.
.TP
.I "\-t"
Tell \fBtack\fR to override the terminfo settings for basic terminal
functions.  When this option is set \fBtack\fR will translate
(cr) to \\r, (cud1) to \\n, (ind) to \\n, (nel) to \\r\\n,
(cub1) to \\b, (bel) to \\007, (ff) to \\f and (ht) to \\t.
.TP
.I "\-V"
Display the version information and exit.
.TP
.I "term"
Terminfo terminal name to be tested.  If not present then the $TERM
environment variable will be used.
.SH OVERVIEW
Since \fBtack\fR is designed to test terminfo's it is not possible
to rely on the correctness of the terminfo data base.  Because of this
the menuing system used with \fBtack\fR is vary primitive.  When a
menu is printed it will scroll the entire screen.  To compensate
for this verbose menu system \fBtack\fR permits menu selection
type ahead.
If you already know what action you would like \fBtack\fR to perform
then you can enter that value immediately and avoid the menu display.
When in doubt the question mark (?) is a good character to type.
A carriage return will execute the default action.  These default
actions are designed to run all the standard tests.
.PP
When \fBtack\fR first comes up it will display some basic information
about the terminal.  Take some time to verify this information.
If it is wrong many of the subsequent tests will fail.  The most
important item is the screen size.  If the screen size is wrong there
is no point in proceeding.  (home) and (clear) are also critical
to the success of subsequent tests.  The values of (cr) (ind)
(cub1) and (ht) may effect the tests if they are defined incorrectly.
If they are undefined \fBtack\fR will set them to reasonable defaults.
The last two entries on the display are the enquire and acknowledge strings.
These strings are taken from the user strings (u9) and (u8).
.PP
By now you must be wondering why the terminfo names are enclosed
in parenthesis.  This has no profound meaning other than it makes
them stand out.  The \fBtack\fR program uses this convention any time
it displays a terminfo name.  Remember \fBtack\fR is designed to
rely on as little of the terminfo entry as possible.
.SH CREATING NEW ENTRIES
\fBTack\fR has a number of tools that are designed to help gather
information about the terminal.  Although these functions are not
dependent on terminal type, you may wish to execute \fBtack\fR
with options \fI\-it\fR.  This will turn off initialization
and default the standard entries.
.PP
These tools may be reached from the main menu by selecting
the 'tools' entry.
.PP
\fBEcho tool\fR:  All data typed from the keyboard will be echoed back
to the terminal.  Control characters are not translated to the up arrow format
but are sent as control characters.  This allows you to test an escape
sequence and see what it actually does.  You may also elect to
\fBenable hex output on echo tool\fR this will echo the characters in
hexadecimal.  Once the test is running you may enter the 'lines'
or 'columns' keywords which will display a pattern that will help
you determine your screen size.  A complete list of keywords will
be displayed when the test starts.  Type 'help' to redisplay
the list of available commands.
.PP
\fBReply tool\fR:  This tool acts much like the echo tool, but
control characters that are sent from the terminal more than one character
after a carriage return will be expanded to the up arrow format.  For example
on a standard ANSI terminal you may type:

		CR ESC [ c

and the response will be echoed as something like:

		^[ [ ? 6 c
.PP
\fBANSI sgr display\fR:  This test assumes you have an ANSI terminal.  It
goes through attribute numbers 0 to 79, displaying each in turn and using that
SGR number to write the text.  This shows you which of the SGR
modes are actually implemented by the terminal.  Note: some terminals (such as
Tektronix color) use the private use characters to augment the functionality of
the SGR command.  These private use characters may be interjected into the
escape sequence by typing the character ( <, =, >, ? ) after the original
display has been shown.
.PP
\fBANSI status reports\fR:  This test queries the terminal in standard
ANSI/VT-100 fashion.  The results of this test may help
determine what options are supported by your terminal.
.PP
\fBANSI character sets\fR:  This test displays the character sets
available on a ANSI/VT-100 style terminal.
Character sets on a real VT-100 terminal are usually defined
with smacs=\\E(0 and rmacs=\\E(B.  The first character after the
escape defines the font bank.  The second character defines the
character set.  This test allows you to view any of the possible
combinations.  Private use character sets are defined by the digits.
Standard character sets are located in the alphabetic range.
.SH VERIFYING AN EXISTING ENTRY
.PP
You can verify the correctness of an entry with the `begin testing'
function.  This entry is the default action and will be chosen
if you hit carriage return (or enter).  This will bring up a
secondary menu that allows you to select more specific tests.
.PP
The general philosophy of the program is, for each capability, to send an
appropriate test pattern to the terminal then send a description of
what the user should expect.  Occasionally (as when checking function-key
capabilities) the program will ask you to enter input for it to check.
.PP
If the test fails then you have the option of dynamically changing
the terminfo entry and re-running the test.  This is done with
the 'edit terminfo' menu item.  The edit submenu allows you to change
the offending terminfo entry and immediately retest the capability.
The edit menu lets you do other things with the terminfo, such as;
display the entire terminfo entry,
display which caps have been tested and display which caps cannot
be tested.  This menu also allows you to write the newly modified
terminfo to disc.  If you have made any modifications to the
terminfo \fBtack\fR will ask you if you want to save the file
to disc before it exits.  The filename will be the same as the terminal name.
After the program exits you can run the tic(1M) compiler on the
new terminfo to install it in the terminfo data base.
.PP
.SH CORRECTING PAD TIMINGS
.SS Theory of Overruns and Padding
.PP
Some terminals require significant amounts of time (that is, more than one
transmitted-character interval) to do screen updates that change large
portions of the screen, such as screen clears, line insertions,
line deletions, and scrolls (including scrolls triggered by line feeds
or a write to the lowest, right-hand-most cell of the screen).
.PP
If the computer continues to send characters to the terminal while one
of these time-consuming operations is going on, the screen may be garbled.
Since the length of a character transmission time varies inversely with
transmission speed in cps, entries which function at lower speeds may
break at higher speeds.
.PP
Similar problems result if the host machine is simply sending characters at a
sustained rate faster than the terminal can buffer and process them.  In either
case, when the terminal cannot process them and can't tell the host to stop
soon enough, it will just drop them.  The dropped characters could be text,
escape sequences or the escape character itself, causing some really
strange-looking displays.  This kind of glitch is called an \fIoverrun\fR.
.PP
In terminfo entries, you can attach a \fBpad time\fR to each string capability
that is a number of milliseconds to delay after sending it.  This will give
the terminal time to catch up and avoid overruns.
.PP
If you are running a software terminal emulator, or you are on an X pseudo-tty,
or your terminal is on an RS-232C line which correctly handles RTS/CTS
hardware flow control, then pads are not strictly necessary.  However, some
display packages (such as ncurses(3X)) use the pad counts to calculate
the fastest way to implement certain functions.
For example: scrolling the screen may be faster than deleting the top line.
.PP
One common way to avoid overruns is with XON/XOFF handshaking.
But even this handshake may have problems at high baud rates.
This is a result of the way XON/XOFF works.  The terminal tells
the host to stop with an XOFF.  When the host gets this character, it stops
sending.  However, there is a small amount of time between the stop request and
the actual stop.  During this window, the terminal must continue to accept
characters even though it has told the host to stop.  If the terminal sends
the stop request too late, then its internal buffer will overflow.  If it sends
the stop character too early, then the terminal is not getting the most
efficient use out of its internal buffers.  In a real application at high baud
rates, a terminal could get a dozen or more characters before the host gets
around to suspending transmission.  Connecting the terminal over a network
will make the problem much worse.
.PP
(RTS/CTS handshaking does not have this problem because the UARTs are
signal-connected and the "stop flow" is done at the lowest level, without
software intervention).
.PP
.SS Timing your terminal
.PP
In order to get accurate timings from your terminal \fBtack\fR 
needs to know when the terminal has finished processing all the
characters that were sent.  This requires a different type of handshaking
than the XON/XOFF that is supported by most terminals.  \fBTack\fR
needs to send a request to the terminal and wait for its reply.
Many terminals will respond with an ACK when they receive an ENQ.
This is the preferred method since the sequence is short.
ANSI/VT-100 style terminals can mimic this handshake with the
escape sequence that requests 'primary device attributes'.

   ESC [ c

The terminal will respond with a sequence like:

   ESC [ ? 1 ; 0 c

\fBTack\fR assumes that (u9) is the enquire sequence and that (u8) is the
acknowledge string.  A VT-100 style terminal could set u9=\\E[c
and u8=\\E[?1;0c.
Acknowledge strings fall into two categories. 
1) Strings with a unique terminating character and,
2) strings of fixed length.
The acknowledge string for the VT-100 is of the first type since
it always ends with the letter 'c'.  Some Tektronics terminals
have fixed length acknowledge strings.  \fBTack\fR supports both
types of strings by scanning for the terminating character until
the length of the expected acknowledge string has arrived.
(u8) should be set to some typical acknowledge that will be
returned when (u9) is sent.
.PP
\fBTack\fR will test this sequence before running any of the pad
tests or the function key tests.  \fBTack\fR will ask you the following:

    Hit lower case g to start testing...

After it sends this message it will send the enquire string.
It will then read characters from the terminal until it sees the
letter g.
.PP
.SS Testing and Repairing Pad Timings
.PP
The pad timings in distributed terminfo entries are often incorrect.  One
major motivation for this program is to make it relatively easy to tune these
timings.
.PP
You can verify and edit the pad timings for a terminal with
the `test string capabilities'
function (this is also part of the `normal test sequence' function).
.PP
The key to determining pad times is to find out the effective baud rate of
the terminal.  The effective baud rate determines the number of characters
per second that the terminal can accept without either handshaking or
losing data.  This rate is frequently less than the nominal cps rate on the
RS-232 line.
.PP
\fBTack\fR uses the effective baud rate to judge the duration of the test and
how much a particular escape sequence will perturb the terminal.
.PP
Each pad test has two associated variables that can be tweaked to help verify
the correctness of the pad timings.  One is the pad test length.  The other is
the pad multiplier, which is used if the pad prefix includes `*'.  In curses
use, it is often the first parameter of the capability (if there is one).
For a capability like (dch) or (il) this will be the number of character
positions or lines affected, respectively.
.PP
\fBTack\fR will run the pad tests and display the results to the terminal.
On capabilities that have multipliers \fBtack\fR will not tell you
if the pad needs the multiplier or not.  You must make this decision
yourself by rerunning the test with a different multiplier.
If the padding changes in proportion to the multiplier than the
multiplier is required.  If the multiplier has little or no effect on
the suggested padding then the multiplier is not needed.
Some capabilities will take several runs to get a good feel for
the correct values.  You may wish to make the test longer
to get more accurate results.  System load will also effect the
results (a heavily loaded system will not stress the
terminal as much, possibly leading to pad timings that are too short).
.PP
.SH NOTE
The tests done at the beginning of the program are assumed to be correct later
in the code.  In particular, \fBtack\fR displays the number of lines and
columns indicated in the terminfo entry as part of its initial output.
If these values are wrong a large number of tests will fail or give incorrect
results.
.SH FILES
.TP 12
tack.log
If logging is enabled then all characters written to the terminal
will also be written to the log file.  This gives you the ability
to see how the tests were performed.  This feature is disabled by default.
.TP 12
.I "term"
If you make changes to the terminfo entry \fBtack\fR will save
the new terminfo to a file.  The file will have the same name
as the terminal name.
.SH SEE ALSO
\fBterminfo\fR(\*n), \fBncurses\fR(3X), \fBtic\fR(1M), \fBinfocmp\fR(1M).
You should also have the documentation supplied by the terminal
manufacturer.
.SH BUGS
If the screen size is incorrect, many of the tests will fail.
.SH AUTHOR
Concept, design, and original implementation by
Daniel Weaver <danw@znyx.com>.  Portions of the code and
documentation are by Eric S. Raymond <esr@snark.thyrsus.com>.
.\"#
.\"# The following sets edit modes for GNU EMACS
.\"# Local Variables:
.\"# mode:nroff
.\"# fill-column:79
.\"# End:


lib/ncurses/tack/tack.c
/*
** Copyright (C) 1991, 1997 Free Software Foundation, Inc.
** 
** This file is part of TACK.
** 
** TACK is free software; you can redistribute it and/or modify
** it under the terms of the GNU General Public License as published by
** the Free Software Foundation; either version 2, or (at your option)
** any later version.
** 
** TACK is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
** GNU General Public License for more details.
** 
** You should have received a copy of the GNU General Public License
** along with TACK; see the file COPYING.  If not, write to
** the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
** Boston, MA 02110-1301, USA
*/

#include <tack.h>

MODULE_ID("$Id$")

/*
   This program is designed to test terminfo, not curses.  Therefore
   I have used as little of curses as possible.

   Pads associated with the following capabilities are used to set
   delay times in the handler:  (cr), (ind), (cub1), (ff), (tab).

   I use the (nxon) capability to set the tty handler with/without
   xon/xoff.  If (smxon)/(rmxon) is defined I will change the terminal
   too.

   (xon) inhibits the sending of delay characters in putp().
   If the terminal is defined with no padding then the (xon) boolean
   is a don't care.  In this case I recommend that it be reset.
 */

/*****************************************************************************
 *
 * Option processing
 *
 *****************************************************************************/

/* options and modes */
int debug_level;		/* debugging level */
int translate_mode;		/* translate tab, bs, cr, lf, ff */
int scan_mode;			/* use scan codes */
int char_mask;			/* either 0xFF else 0x7F, eight bit data mask */
int select_delay_type;		/* set handler delays for <cr><lf> */
int select_xon_xoff;		/* TTY driver XON/XOFF mode select */
int hex_out;			/* Display output in hex */
int send_reset_init;		/* Send the reset and initialization strings */
FILE *log_fp;			/* Terminal logfile */

/*****************************************************************************
 *
 * Menu definitions
 *
 *****************************************************************************/

static void tools_hex_echo(struct test_list *, int *, int *);
static void tools_debug(struct test_list *, int *, int *);

static char hex_echo_menu_entry[80];

static struct test_list tools_test_list[] = {
	{0, 0, 0, 0, "s) ANSI status reports", tools_status, 0},
	{0, 0, 0, 0, "g) ANSI SGR modes (bold, underline, reverse)", tools_sgr, 0},
	{0, 0, 0, 0, "c) ANSI character sets", tools_charset, 0},
	{0, 0, 0, 0, hex_echo_menu_entry, tools_hex_echo, 0},
	{0, 0, 0, 0, "e) echo tool", tools_report, 0},
	{1, 0, 0, 0, "r) reply tool", tools_report, 0},
	{0, 0, 0, 0, "p) performance testing", 0, &sync_menu},
	{0, 0, 0, 0, "i) send reset and init", menu_reset_init, 0},
	{0, 0, "u8) (u9", 0, "u) test ENQ/ACK handshake", sync_handshake, 0},
	{0, 0, 0, 0, "d) change debug level", tools_debug, 0},
	{MENU_LAST, 0, 0, 0, 0, 0, 0}
};

static struct test_menu tools_menu = {
	0, 'q', 0, "Tools Menu", "tools",
	0, 0, tools_test_list, 0, 0, 0
};

static void tty_width(struct test_list *, int *, int *);
static void tty_delay(struct test_list *, int *, int *);
static void tty_xon(struct test_list *, int *, int *);
static void tty_trans(struct test_list *, int *, int *);
static void tty_show_state(struct test_menu *);

static char tty_width_menu[80];
static char tty_delay_menu[80];
static char tty_xon_menu[80];
static char tty_trans_menu[80];
static char enable_xon_xoff[] = {"x) enable xon/xoff"};
static char disable_xon_xoff[] = {"x) disable xon/xoff"};

static struct test_list tty_test_list[] = {
	{0, 0, 0, 0, tty_width_menu, tty_width, 0},
	{0, 0, 0, 0, tty_delay_menu, tty_delay, 0},
	{0, 0, 0, 0, tty_xon_menu, tty_xon, 0},
	{0, 0, 0, 0, tty_trans_menu, tty_trans, 0},
	{MENU_LAST, 0, 0, 0, 0, 0, 0}
};

static struct test_menu tty_menu = {
	0, 'q', 0, "Terminal and driver configuration",
	"tty", 0,
	tty_show_state, tty_test_list, 0, 0, 0
};

struct test_menu edit_menu = {
	0, 'q', 0, "Edit terminfo menu",
	"edit", 0,
	0, edit_test_list, 0, 0, 0
};

static struct test_menu mode_menu = {
	0, 'n', 0, "Test modes and glitches:",
	"mode", "n) run standard tests",
	0, mode_test_list, 0, 0, 0
};

static struct test_menu acs_menu = {
	0, 'n', 0,
	"Test alternate character set and graphics rendition:",
	"acs", "n) run standard tests",
	0, acs_test_list, 0, 0, 0
};

static struct test_menu color_menu = {
	0, 'n', 0,
	"Test color:",
	"color", "n) run standard tests",
	0, color_test_list, 0, 0, 0
};

static struct test_menu crum_menu = {
	0, 'n', 0,
	"Test cursor movement:",
	"move", "n) run standard tests",
	0, crum_test_list, 0, 0, 0
};

static struct test_menu funkey_menu = {
	0, 'n', 0,
	"Test function keys:",
	"fkey", "n) run standard tests",
	sync_test, funkey_test_list, 0, 0, 0
};

static struct test_menu printer_menu = {
	0, 'n', 0,
	"Test printer:",
	"printer", "n) run standard tests",
	0, printer_test_list, 0, 0, 0
};

static void pad_gen(struct test_list *, int *, int *);

static struct test_menu pad_menu = {
	0, 'n', 0,
	"Test padding and string capabilities:",
	"pad", "n) run standard tests",
	sync_test, pad_test_list, 0, 0, 0
};

static struct test_list normal_test_list[] = {
	{0, 0, 0, 0, "e) edit terminfo", 0, &edit_menu},
	{0, 0, 0, 0, "i) send reset and init", menu_reset_init, 0},
	{MENU_NEXT, 0, 0, 0, "x) test modes and glitches", 0, &mode_menu},
	{MENU_NEXT, 0, 0, 0, "a) test alternate character set and graphic rendition", 0, &acs_menu},
	{MENU_NEXT, 0, 0, 0, "c) test color", 0, &color_menu},
	{MENU_NEXT, 0, 0, 0, "m) test cursor movement", 0, &crum_menu},
	{MENU_NEXT, 0, 0, 0, "f) test function keys", 0, &funkey_menu},
	{MENU_NEXT, 0, 0, 0, "p) test padding and string capabilities", 0, &pad_menu},
	{0, 0, 0, 0, "P) test printer", 0, &printer_menu},
	{MENU_MENU, 0, 0, 0, "/) test a specific capability", 0, 0},
	{0, 0, 0, 0, "t) auto generate pad delays", pad_gen, &pad_menu},
	{0, 0, "u8) (u9", 0, 0, sync_handshake, 0},
	{MENU_LAST, 0, 0, 0, 0, 0, 0}
};


static struct test_menu normal_menu = {
	0, 'n', 0, "Main test menu",
	"test", "n) run standard tests",
	0, normal_test_list, 0, 0, 0
};

static void start_tools(struct test_list *, int *, int *);
static void start_modes(struct test_list *, int *, int *);
static void start_basic(struct test_list *, int *, int *);
static void start_log(struct test_list *, int *, int *);

static char logging_menu_entry[80] = "l) start logging";

static struct test_list start_test_list[] = {
	{0, 0, 0, 0, "b) display basic information", start_basic, 0},
	{0, 0, 0, 0, "m) change modes", start_modes, 0},
	{0, 0, 0, 0, "t) tools", start_tools, 0},
	{MENU_COMPLETE, 0, 0, 0, "n) begin testing", 0, &normal_menu},
	{0, 0, 0, 0, logging_menu_entry, start_log, 0},
	{MENU_LAST, 0, 0, 0, 0, 0, 0}
};
	

static struct test_menu start_menu = {
	0, 'n', 0, "Main Menu", "tack", 0,
	0, start_test_list, 0, 0, 0
};

static struct test_list write_terminfo_list[] = {
	{0, 0, 0, 0, "w) write the current terminfo to a file", save_info, 0},
	{MENU_LAST, 0, 0, 0, 0, 0, 0}
};

/*****************************************************************************
 *
 * Menu command interpretation.
 *
 *****************************************************************************/

/*
**	tools_hex_echo(testlist, state, ch)
**
**	Flip the hex echo flag.
*/
static void
tools_hex_echo(
	struct test_list *t GCC_UNUSED,
	int *state GCC_UNUSED,
	int *ch GCC_UNUSED)
{
	if (hex_out) {
		hex_out = FALSE;
		strcpy(hex_echo_menu_entry,
			"h) enable hex output on echo tool");
	} else {
		hex_out = TRUE;
		strcpy(hex_echo_menu_entry,
			"h) disable hex output on echo tool");
	}
}

/*
**	tools_debug(testlist, state, ch)
**
**	Change the debug level.
*/
static void
tools_debug(
	struct test_list *t GCC_UNUSED,
	int *state GCC_UNUSED,
	int *ch)
{
	char buf[32];

	ptext("Enter a new value: ");
	read_string(buf, sizeof(buf));
	if (buf[0]) {
		sscanf(buf, "%d", &debug_level);
	}
	sprintf(temp, "Debug level is now %d", debug_level);
	ptext(temp);
	*ch = REQUEST_PROMPT;
}

/*
**	start_tools(testlist, state, ch)
**
**	Run the generic test tools
*/
static void
start_tools(
	struct test_list *t GCC_UNUSED,
	int *state GCC_UNUSED,
	int *ch GCC_UNUSED)
{
	if (hex_out) {
		strcpy(hex_echo_menu_entry,
			"h) disable hex output on echo tool");
	} else {
		strcpy(hex_echo_menu_entry,
			"h) enable hex output on echo tool");
	}
	menu_display(&tools_menu, 0);
}

/*
**	tty_show_state()
**
**	Display the current state on the tty driver settings
*/
static void
tty_show_state(
	struct test_menu *menu GCC_UNUSED)
{
	put_crlf();
	(void) sprintf(temp,
		"Accepting %d bits, UNIX delays %d, XON/XOFF %sabled, speed %u, translate %s, scan-code mode %s.",
		(char_mask == ALLOW_PARITY) ? 8 : 7,
		select_delay_type,
		select_xon_xoff ? "en" : "dis",
		tty_baud_rate,
		translate_mode ? "on" : "off",
		scan_mode ? "on" : "off");
	ptextln(temp);
	put_crlf();
}

/*
**	tty_width(testlist, state, ch)
**
**	Change the character width
*/
static void
tty_width(
	struct test_list *t GCC_UNUSED,
	int *state GCC_UNUSED,
	int *ch GCC_UNUSED)
{
	if (char_mask == STRIP_PARITY) {
		char_mask = ALLOW_PARITY;
		strcpy(tty_width_menu, "7) treat terminal as 7-bit");
	} else {
		char_mask = STRIP_PARITY;
		strcpy(tty_width_menu, "8) treat terminal as 8-bit");
	}
}

/*
**	tty_delay(testlist, state, ch)
**
**	Change the delay for <cr><lf> in the TTY driver
*/
static void
tty_delay(
	struct test_list *t GCC_UNUSED,
	int *state GCC_UNUSED,
	int *ch GCC_UNUSED)
{
	if (select_delay_type) {
		select_delay_type = FALSE;
		strcpy(tty_delay_menu,
			"d) enable UNIX tty driver delays for <cr><lf>");
	} else {
		select_delay_type = TRUE;
		strcpy(tty_delay_menu,
			"d) disable UNIX tty driver delays for <cr><lf>");
	}
}

/*
**	tty_xon(testlist, state, ch)
**
**	Change the XON/XOFF flags in the TTY driver
*/
static void
tty_xon(
	struct test_list *t GCC_UNUSED,
	int *state GCC_UNUSED,
	int *ch GCC_UNUSED)
{
	if (select_xon_xoff) {
		if (needs_xon_xoff) {
			ptextln("This terminal is marked as needing XON/XOFF protocol with (nxon)");
		}
		if (exit_xon_mode) {
			tc_putp(exit_xon_mode);
		}
		xon_xoff = select_xon_xoff = FALSE;
		strcpy(tty_xon_menu, enable_xon_xoff);
	} else {
		if (enter_xon_mode) {
			tc_putp(enter_xon_mode);
		}
		xon_xoff = select_xon_xoff = TRUE;
		strcpy(tty_xon_menu, disable_xon_xoff);
	}
	tty_set();
}

/*
**	tty_trans(testlist, state, ch)
**
**	Change the translation mode for special characters
*/
static void
tty_trans(
	struct test_list *t GCC_UNUSED,
	int *state GCC_UNUSED,
	int *ch GCC_UNUSED)
{
	if (translate_mode) {
		translate_mode = FALSE;
		strcpy(tty_trans_menu,
			"t) use terminfo values for \\b\\f\\n\\r\\t");
	} else {
		translate_mode = TRUE;
		strcpy(tty_trans_menu,
			"t) override terminfo values for \\b\\f\\n\\r\\t");
	}
}

/*
**	pad_gen(testlist, state, ch)
**
**	Menu function for automatic pad generation
*/
static void
pad_gen(
	struct test_list *t,
	int *state GCC_UNUSED,
	int *ch)
{
	control_init();
	if (tty_can_sync == SYNC_NOT_TESTED) {
		verify_time();
	}
	auto_pad_mode = TRUE;
	menu_display(t->sub_menu, ch);
	auto_pad_mode = FALSE;
}

/*
**	start_modes(testlist, state, ch)
**
**	Change the TTY modes
*/
static void
start_modes(
	struct test_list *t GCC_UNUSED,
	int *state GCC_UNUSED,
	int *ch GCC_UNUSED)
{

	if (select_delay_type) {
		strcpy(tty_delay_menu,
			"d) disable UNIX tty driver delays for <cr><lf>");
	} else {
		strcpy(tty_delay_menu,
			"d) enable UNIX tty driver delays for <cr><lf>");
	}
	if (char_mask == ALLOW_PARITY) {
		strcpy(tty_width_menu,
			"7) treat terminal as 7-bit");
	} else {
		strcpy(tty_width_menu,
			"8) treat terminal as 8-bit");
	}
	if (select_xon_xoff) {
		strcpy(tty_xon_menu, disable_xon_xoff);
	} else {
		strcpy(tty_xon_menu, enable_xon_xoff);
	}
	if (translate_mode) {
		strcpy(tty_trans_menu,
			"t) override terminfo values for \\b\\f\\n\\r\\t");
	} else {
		strcpy(tty_trans_menu,
			"t) use terminfo values for \\b\\f\\n\\r\\t");
	}
	menu_display(&tty_menu, 0);
	tty_set();
}

/*
**	start_basic(testlist, state, ch)
**
**	Display basic terminal information
*/
static void
start_basic(
	struct test_list *t GCC_UNUSED,
	int *state GCC_UNUSED,
	int *ch)
{
	display_basic();
	*ch = REQUEST_PROMPT;
}

/*
**	start_log(testlist, state, ch)
**
**	Start/stop in logging function
*/
static void
start_log(
	struct test_list *t GCC_UNUSED,
	int *state GCC_UNUSED,
	int *ch GCC_UNUSED)
{
	if (logging_menu_entry[5] == 'a') {
		ptextln("The log file will capture all characters sent to the terminal.");
		if ((log_fp = fopen("tack.log", "w"))) {
			ptextln("Start logging to file: tack.log");
			strcpy(logging_menu_entry, "l) stop logging");
		} else {
			ptextln("File open error: tack.log");
		}
	} else {
		if (log_fp) {
			fclose(log_fp);
			log_fp = 0;
		}
		ptextln("Terminal output logging stopped.");
		strcpy(logging_menu_entry, "l) start logging");
	}
}

/*
**	show_usage()
**
**	Tell the user how its done.
*/
void
show_usage(
	char *name)
{
	(void) fprintf(stderr, "usage: %s [-itV] [term]\n", name);
}

/*
**	print_version()
**
**	Print version and other useful information.
*/
void
print_version(void)
{
	printf("tack version %d.%02d\n", MAJOR_VERSION, MINOR_VERSION);
	printf("Copyright (C) 1997 Free Software Foundation, Inc.\n");
	printf("Tack comes with NO WARRANTY, to the extent permitted by law.\n");
	printf("You may redistribute copies of Tack under the terms of the\n");
	printf("GNU General Public License.  For more information about\n");
	printf("these matters, see the file named COPYING.\n");
}


/*****************************************************************************
 *
 * Main sequence
 *
 *****************************************************************************/

int
main(int argc, char *argv[])
{
	int i, j;
	char *term_variable;

	/* scan the option flags */
	send_reset_init = TRUE;
	translate_mode = FALSE;
	term_variable = getenv("TERM");
	tty_can_sync = SYNC_NOT_TESTED;
	for (i = 1; i < argc; i++) {
		if (argv[i][0] == '-') {
			for (j = 1; argv[i][j]; j++) {
				switch (argv[i][j]) {
				case 'V':
					print_version();
					return (1);
				case 'i':
					send_reset_init = FALSE;
					break;
				case 't':
					translate_mode = FALSE;
					break;
				default:
					show_usage(argv[0]);
					return (0);
				}
			}
		} else {
			term_variable = argv[i];
		}
	}
	(void) strcpy(tty_basename, term_variable);

	curses_setup(argv[0]);

	menu_can_scan(&normal_menu);	/* extract which caps can be tested */
	menu_display(&start_menu, 0);

	if (user_modified()) {
		sprintf(temp, "Hit y to save changes to file: %s  ? ",
			tty_basename);
		ptext(temp);
		if (wait_here() == 'y') {
			save_info(write_terminfo_list, &i, &j);
		}
	}

	put_str("\nTerminal test complete\n");
	bye_kids(0);
	return (0);
}
